client/my-sites/checkout/checkout/web-payment-box.jsx

```diff
@@ -28,523 +26,383 @@ import {
+		} ),
+		[ translate, countryCode, currency, total_cost_integer, displayItems, shouldDisplayItems ]
+	);
+	return paymentRequestOptions;
+}
 
-		if ( 'postal-code' === key ) {
-			setTaxPostalCode( value );
+function useStripePaymentRequest( {
+	stripe,
+	paymentRequestOptions,
+	countryCode,
+	postalCode,
+	onSubmit,
+	isRenewal,
+} ) {
+	const [ canMakePayment, setCanMakePayment ] = useState( false );
+	const [ paymentRequest, setPaymentRequest ] = useState();
+
+	// We have to memoize this to prevent re-creating the paymentRequest
+	const callback = useMemo(
+		() => paymentMethodResponse => {
+			analytics.tracks.recordEvent( 'calypso_checkout_apple_pay_submit_payment_sheet', {
+				is_renewal: isRenewal,
+			} );
+			completePaymentMethodTransaction( {
+				countryCode,
+				postalCode,
+				onSubmit,
+				...paymentMethodResponse,
+			} );
+		},
+		[ countryCode, postalCode, onSubmit, isRenewal ]
+	);
+
+	useEffect( () => {
+		let isSubscribed = true;
+		if ( ! stripe ) {
+			return;
 		}
-	};
-
-	render() {
-		/* eslint-disable wpcalypso/jsx-classname-namespace */
-		const paymentMethod = this.detectedPaymentMethod;
-		const { cart, countriesList, presaleChatAvailable, translate } = this.props;
-		const countryCode = getTaxCountryCode( cart );
-		const postalCode = getTaxPostalCode( cart );
+		debug( 'creating paymentRequest', paymentRequestOptions );
+		setCanMakePayment( false );
+		const request = stripe.paymentRequest( paymentRequestOptions );
+		request.canMakePayment().then( result => {
+			isSubscribed && setCanMakePayment( !! result );
+		} );
+		request.on( 'paymentmethod', callback );
+		setPaymentRequest( request );
+		return () => ( isSubscribed = false );
+	}, [ stripe, paymentRequestOptions, callback ] );
+
+	return { paymentRequest, canMakePayment };
+}
 
-		if ( ! paymentMethod ) {
-			return null;
-		}
+function LoadingPaymentRequestButton( { paymentType, translate } ) {
+	if ( paymentType === 'apple-pay' ) {
+		return <button className="web-payment-box__apple-pay-button" disabled />;
+	}
+	return (
+		<button
+			className="web-payment-box__web-pay-button button checkout__pay-button-button button is-primary button-pay pay-button__button"
+			disabled
+		>
+			{ translate( 'Select a payment card' ) }
+		</button>
+	);
+}
 
-		const hasBusinessPlanInCart = some( cart.products, ( { product_slug } ) =>
-			overSome( isWpComBusinessPlan, isWpComEcommercePlan )( product_slug )
-		);
-		const showPaymentChatButton = presaleChatAvailable && hasBusinessPlanInCart;
-
-		const buttonState = this.getButtonState();
-		const buttonDisabled = buttonState.disabled;
-		let button;
-		let paymentType;
-
-		switch ( paymentMethod ) {
-			case WEB_PAYMENT_APPLE_PAY_METHOD:
-				if ( buttonState.default ) {
-					button = (
-						<button
-							type="submit"
-							onClick={ event =>
-								! countryCode || ! postalCode
-									? this.displaySubmissionError(
-											translate( 'Please specify a country and postal code.' ),
```

Comment:
Is there a way to keep this?

The user experience without it (where the button is just grayed out with no indication of why it can't be submitted) definitely seems much worse.

Reply:
Good point. I'll make something.

Reply:
Done.